cocotb/regression.py

```diff
@@ -641,7 +641,9 @@ class TestFactory:
     """
 
     def __init__(self, test_function, *args, **kwargs):
-        if not isinstance(test_function, cocotb.coroutine):
+        if inspect.isasyncgenfunction(test_function):
+            raise TypeError("Found yield in async function '%s'" % test_function.__name__)
```

Comment:
Apologies, was on mobile earlier so had difficultly suggesting a message
```suggestion
        if sys.version_info > (3, 6) and inspect.isasyncgenfunction(test_function):
            raise TypeError(
                "Expected a coroutine function, but got the async generator '{}'. "
                "Did you forget to convert a `yield` to an `await`?"
                .format(test_function.__qualname__))
            )
```
Also needs a version check since `isasyncgenfunction` is new in 3.6.